### ISSUE TITLE
Detect when a failure might be due to needing a build

### ISSUE THREAD

**Issue by garyb** [kRITZCREEK/pscid#8]
This isn't something for the near future necessarily, but it would be nice if we could do something cunning where we detect that an error is raised in a file that is a dependant of a previously edited file since the last build was triggered, and suggest that a full build may help solve the problem. Perhaps restricted to when it's a type error / unknown value/module, etc.

**Comment by kritzcreek** [comment]
I think we should put such cunningness into psc-ide-server so that all the tools can benefit from it.

**Comment by garyb** [comment]
Yeah agreed. And actually after a bit more time with `pscid` the cases I had in mind are don't require a full rebuild, just a reset of `psc-ide-server`.

**Comment by kritzcreek** [comment]
Can we detect these easily? In that case we could point the user to the `r` binding.

**Comment by garyb** [comment]
It's probably easier to do it in `psc-ide-server` actually as you'd need to do something like this:
- hold module graph in memory
- track which modules have been modified
- observe when a module (Y) is edited that depends on another edited module (X)

This last case is where the reset needs triggering, as if X has had members added or something like that, Y will not be able to refer to them until `psc-ide-server` is reset.

From some of the discussion I saw in passing (I wasn't following the PR too closely I must admit) I thought this work was already done by psc-ide for the most part, just not the auto-reset thing. I guess it wouldn't need a reset as such, just an update of the environment or something, I'm not exactly sure what it is internally that leads to this.

**Comment by kritzcreek** [comment]
I've seen this aswell, and I think the problem lies within the reliability of the file watcher in Haskell, but I'm not sure and it would be best to have a reproducible testcase before starting on it. Anyway out of scope for pscid.

**Comment by garyb** [comment]
Ah ok, interesting. It seemed to happen every time I attempted to do something like the above - I'll try and put a test case together at the main repo.

**Comment by kritzcreek** [comment]
That would be great!
